### ISSUE TITLE
Decompiler does not raise non-capturing void-body (Action) lambdas → leaks <>c names, invalid C# (CS1525)

### ISSUE THREAD

**Issue by richlander** [richlander/dotnet-inspect#1617]
## Product area / surface

`ILInspector.Decompiler` lambda raising (`LambdaRaisingPass` /
`LambdaCachePass`) and the C# body printer. A **non-capturing lambda whose
delegate return type is `void` (an `Action`-family lambda)** is not raised back to
`=>` syntax; the fallback prints the cached delegate creation with the
compiler-generated closure members, e.g. `new Action(<>c.<>9.<M>b__N_M)`, which
is not valid C#. Value-returning (`Func`) non-capturing lambdas raise correctly.

## False claim being falsified

`ClosureCoverage.Lambda` is marked `Partial` claiming
*"expression, simple block, non-capturing local-bound bodies, and
parameter-capturing local-bound bodies"* are recovered; only
*"outer-local capturing local-bound bodies"* are owed. That claim does not carve
out `void`-bodied lambdas, yet every non-capturing `Action` lambda falls through
unrecovered and renders invalid C# (CS1525), while the equivalent `Func` lambda
recovers. So the "non-capturing ... recovered" claim is false for `void` bodies.

## Minimal repro (ordinary source)

```csharp
public static class Probe
{
    public static Func<int>      ParamlessFunc() => () => 5;                 // recovered
    public static Func<int,int>  FuncBlock()     => x => { return x + 1; };  // recovered
    public static Action         EmptyBody()     => () => { };               // NOT recovered
    public static Action         VoidExpr()      => () => GC.Collect();      // NOT recovered
    public static Action<int>    ActionParam()   => i => Console.WriteLine(i); // NOT recovered
    public static Action         RetViaLocal()   { Action a = () => Console.WriteLine("hi"); return a; } // NOT recovered
}
```

## Observed bad output

| Method | Delegate | Rendered C# | Result |
| --- | --- | --- | --- |
| `ParamlessFunc` | `Func<int>` | `() => 5` | OK |
| `FuncBlock` | `Func<int,int>` | `x => x + 1` | OK |
| `EmptyBody` | `Action` | `new Action(<>c.<>9.<EmptyBody>b__2_0)` | CS1525 |
| `VoidExpr` | `Action` | `new Action(<>c.<>9.<VoidExpr>b__4_0)` | CS1525 |
| `ActionParam` | `Action<int>` | `new Action<int>(<>c.<>9.<ActionParam>b__1_0)` | CS1525 |
| `RetViaLocal` | `Action` | `new Action(<>c.<>9.<RetViaLocal>b__2_0)` | CS1525 |

Compile proof:

```
error CS1525: Invalid expression term '<'
error CS1525: Invalid expression term '>'
error CS1001: Identifier expected
```

The discriminator is purely the delegate's return type: `Func`/value-bodied
lambdas raise; `Action`/`void`-bodied lambdas do not. It is independent of arity
(`() =>` and `i =>` both fail) and of escaping vs local-bound (direct return and
return-via-local both fail).

## Root cause

`LambdaRaisingPass` recovers lambdas through the value-producing (return-value)
path; a `void`-bodied lambda has no return value — its body is an
expression-statement — so it is not matched and falls through to the raw
`new Action(methodgroup)` printer, which emits the compiler-generated
`<>c.<>9.<M>b__N_M` members verbatim (invalid identifiers).

## First fix shape

Extend the non-capturing lambda recovery to `void`-bodied (`Action`-family)
lambdas: raise the cached `<>c`-hosted `b__N_M` method whose signature returns
`void` into a statement-bodied lambda (`() => Console.WriteLine("hi")`,
`() => { }`, `i => Console.WriteLine(i)`). If a void-body shape cannot be raised,
degrade to `Partial` with a diagnostic instead of emitting `<>c.*` mangled names
as `Full`. Update the `ClosureCoverage.Lambda` completeness text to match
whatever boundary actually holds.

## Done signal

- `EmptyBody` -> `() => { }`, `VoidExpr` -> `() => GC.Collect()`,
  `ActionParam` -> `i => Console.WriteLine(i)`, `RetViaLocal` ->
  `() => Console.WriteLine("hi")`, all compiling (no CS1525), or reported
  `Partial` with a diagnostic and no leaked `<>c` names.
- Positive canaries that must keep working: `ParamlessFunc` -> `() => 5`,
  `FuncBlock` -> `x => x + 1`, `Func` lambdas passed to LINQ (`src.Where(x => x > 0)`),
  and method-group conversions (`new Action(Probe.S)`).
- Add focused lambda-raising tests for void-bodied non-capturing lambdas;
  validate with `dotnet run --project src/ILInspector.Decompiler.Tests -c Release`.

## Notes

- `Action`/void-returning delegates are extremely common (callbacks, event
  handlers, `Task.Run(() => ...)`), so corpus impact is real.
- Leaking `<>c`/`b__N_M` identifiers as `Full` violates the "honest degradation"
  floor — these names are never valid C#.
- No existing open issue covers this (searched open+closed).

**Comment by richlander** [comment]
Closing as author after verifying against current main (05eea338) with the harness.

**This is not a Full-validity defect — it is honest `Partial` degradation.** `DecompilerHarness --dump 'Probe::ActLam' --remarks` reports:

```
fidelity: Partial
remarks: 2 sites cap fidelity below Full
  DEC0009  IL_0000  type name '<>c' has no C# spelling
```

The unrecovered non-capturing void (`Action`) lambda is correctly marked `Partial` with `DEC0009` (the `<>c` closure type has no C# spelling). It is not claimed `Full`. Under the project quality model, `Partial`-malformed output is expected, so the leaked `<>c.<>9.<M>b__N_M` text is documented honest degradation.

My 'renders Full / falsifies the non-capturing-recovered claim' framing was inaccurate: the `ClosureCoverage.Lambda` text scopes recovery to *local-bound* bodies, and these escaping/return-position `Action` lambdas are an owed raise gap that is honestly degraded, not a Full bug. (`Func`/value lambdas do recover at `Full`.) The remaining value is a *raise improvement* for escaping void lambdas, not a validity defect. Closing.
